### ISSUE TITLE
Unable to run megaqc upload from docker container

### ISSUE THREAD

**Issue by jethror1** [ewels/MegaQC#324]
When running megaQC in a docker container with docker run as per the docs [here](https://megaqc.info/docs/installation/installation.html#running-megaqc-with-a-local-postgres-database), it doesn't seem possible to pass in the config file required for using `megaqc upload`. Not sure if I'm just missing something obvious or if it's an issue with megaQC .

We have megaQC set up with a local installation of postgres and everything else seems to be working, but it isn't clear how to pass the config file. Even if we use `docker exec -it` to go into the container and pass the config file with url and token where the code looks to expect it, it always will just print no URL found:

Copied `megaqc_config.yaml` to megaqc dir as it looks to be able to find it here: https://github.com/ewels/MegaQC/blob/27d6619bb1616614a19e59f9bad41496d53049f7/megaqc/utils/settings.py#L61

```console
$ docker exec -it d070b9988d28 ls megaqc
__init__.py  commands.py         migrations   scheduler.py  utils
__pycache__  compat.py           model        settings.py   wsgi.py
api          database.py         public       static
app.py       extensions.py       report_plot  templates
cli.py       megaqc_config.yaml  rest_api     user
```

Output from running `upload`:

```console
$ docker exec -it d070b9988d28 megaqc upload /uploads/multiqc_data.json
This is MegaQC v0.3.0

 * Environment: Prod
 * Database type: postgresql
 * Database path: postgresql://megaqc:***@127.0.0.1:5432/megaqc
Error - MultiQC megaqc_url not set.
```

Contents of `megaqc_config.yaml` (copied with megaqc in browser):

```yaml
megaqc_url = http://127.0.0.1/api/upload_data
megaqc_access_token = [redacted]
```

Also tried including a config named `.multiqc_config.yaml` by binding a dir containing it to `:/root/` in the container which also didn't work.

**To Reproduce**

Steps to reproduce the behaviour:

`sudo docker run -v /appdata/megaqc:/uploads:z -v $(pwd):/root:z --network=”host” ewels/megaqc:chris`

- `/appdata/megaqc`: multiqc data json location
- `$(pwd)` : mounted to `/root` containing `.megaqc_config.yaml`

**Expected behavior**

`docker upload` runs successfully. 

**Additional context**

As an aside, could the URL and access token be also pulled from environment variables? Then when running the docker image you could pass `--env megaqc_url=URL` and `--env megaqc_access_token=TOKEN` and not need to pass in the config file.

Thanks!

**Comment by ewels** [comment]
Could it be that the YAML config file is malformed? Should be `:` and not `=`

```yaml
megaqc_url: http://127.0.0.1/api/upload_data
megaqc_access_token: [redacted]
```

**Comment by multimeric** [comment]
Also, `megaqc upload` is a bit of an unusual command because it POSTS a multiqc report to another URL which isn't even necessarily the same instance of MegaQC. So it would make more sense to clone MegaQC locally, setup the config file locally to point at the URL of the docker container, and then run `megaqc upload` locally.

But actually it's likely much simpler to just POST to [`/rest_api/v1/uploads`](https://megaqc.info/api_reference.html#post--rest_api-v1-uploads) because that has no local dependencies. You can use cURL or postman or whatever.

**Comment by jethror1** [comment]
> Could it be that the YAML config file is malformed? Should be `:` and not `=`
> 
> ```yaml
> megaqc_url: http://127.0.0.1/api/upload_data
> megaqc_access_token: [redacted]
> ```

Apologies I meant `:`, was trying to remember from Friday afternoon (obviously not very well) how it was configured without going back on the server.

@multimeric Using a POST request would probably be easier, will take a look at this thanks.

I read this in the docs: https://megaqc.info/docs/installation/installation.html#setup-your-access-key and did just try rerun it by naming the config file `multiqc_config.yaml` instead of `megaqc_config.yaml`, and running it from `/app/megaqc` inside the container and it seems to have worked.

This seems very confusing as I believe the missing URL message is coming from here: https://github.com/ewels/MegaQC/blob/27d6619bb1616614a19e59f9bad41496d53049f7/megaqc/commands.py#L199

But then `megaqc_load_userconfig()` expects it to be called `megaqc_config.yaml` here: https://github.com/ewels/MegaQC/blob/27d6619bb1616614a19e59f9bad41496d53049f7/megaqc/utils/settings.py#L61

Is this because the code in the docker image is different to that in the main branch on github?

**Comment by ewels** [comment]
hah, yes this is very confusing and probably my fault (sorry). The access key should be set in `multiqc_config.yml` because usually it's MultiQC that sends data to MegaQC, at run time.

In the early days of MegaQC I tried to emulate the MultiQC configuration mechanism which is where the `megaqc_config.yaml` stuff comes from. But I have a feeling that in the end this particular chunk of code actually does nothing (?) and we use the normal Flask route of [environment variables](https://megaqc.info/docs/installation/installation.html#export-environment-variables) etc. for configuration of MegaQC.

So yeah, the old yaml config configuration stuff should probably be cleaned up and removed.

**Comment by ewels** [comment]
ps. Not entirely sure if the `megaqc upload` command line tool uses `megaqc_config.yml` or what. Sorry, been too long since I wrote / used this myself.

**Comment by jethror1** [comment]
No worries, thanks for the explanation. We have something now working so this isn't an issue, probably will have more questions when we get into using megaqc more, but I'm happy to close this issue.

Are you happy for contributions to megaqc? And if so would you prefer issues opening first to discuss how you would like features / bugs addressing (or if you're already working on them), or to just open PRs for you to review?

There's a few minor things I've found and would be happy to work on them when I've time and open PRs instead of adding more to your workload.

Thanks again for megaQC and all your help :+1:

**Comment by multimeric** [comment]
Great! How did you end up uploading the reports?

Contributions are very welcome, I think it's best to open an issue or find a relevant one first. I can probably point you in the right direction.

**Comment by jethror1** [comment]
We have all of our runs and multiqc data in AWS, so we have written a script to periodically check for new data and pull it down, then just call megaqc upload on the downloaded JSON file(s). For now I have just built our own docker image with this script included and it seems to work as intended, I did try just posting to the upload endpoint but ran into firewall issues with our local webserver setup so I'm happy with it running from the same image for now.

And thanks for that, will open issues as I come across anything that I'm not sure about.
